File: LeetCode/P503_NextGreaterElementII.java
/*
 * LeetCode 503 - Next Greater Element II
 * https://leetcode.com/problems/next-greater-element-ii/
 *
 * Approach: Monotonic decreasing stack, iterate over the array twice
 *           (2n loop) to simulate circular behavior.
 * Time  : O(n)
 * Space : O(n)
 */

import java.util.*;

public class P503_NextGreaterElementII {

    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];
        Arrays.fill(result, -1);
        Deque<Integer> stack = new ArrayDeque<>(); // stores indices

        for (int i = 0; i < 2 * n; i++) {
            int num = nums[i % n];
            while (!stack.isEmpty() && nums[stack.peek()] < num) {
                result[stack.pop()] = num;
            }
            if (i < n) {
                stack.push(i);
            }
        }
        return result;
    }

    // ---------- Simple manual test ----------
    public static void main(String[] args) {
        P503_NextGreaterElementII sol = new P503_NextGreaterElementII();

        int[] nums = {1, 2, 1};
        System.out.println(Arrays.toString(sol.nextGreaterElements(nums))); // [2, -1, 2]
    }
}

File: LeetCode/P503_NextGreaterElementII.cpp
#include <bits/stdc++.h>
using namespace std;

class P503_NextGreaterElementII {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = (int) nums.size();
        vector<int> result(n, -1);
        stack<int> stk;

        for (int i = 0; i < 2 * n; i++) {
            int num = nums[i % n];
            while (!stk.empty() && nums[stk.top()] < num) {
                result[stk.top()] = num;
                stk.pop();
            }
            if (i < n) {
                stk.push(i);
            }
        }
        return result;
    }
};

string toString(const vector<int>& arr) {
    ostringstream sb;
    sb << "[";
    for (size_t i = 0; i < arr.size(); i++) {
        if (i > 0) {
            sb << ", ";
        }
        sb << arr[i];
    }
    sb << "]";
    return sb.str();
}

int main() {
    P503_NextGreaterElementII sol;

    vector<int> nums = {1, 2, 1};
    cout << toString(sol.nextGreaterElements(nums)) << "\n";

    return 0;
}
